### ISSUE TITLE
VoiceOver x Safari issues on action menu

### ISSUE THREAD

**Issue by joshkimux** [adobe/react-spectrum#8998]
### Provide a general summary of the issue here

When using the [react aria action menu](https://react-spectrum.adobe.com/react-aria/examples/action-menu.html#action-menu) with VoiceOver and Safari, there are multiple issues that emerge including:
- Broken navigation
- Menu autofocuses the first menu item, but fails to announce it
- Menu incorrectly announces number of menu items

Replication steps will be documented in the current behavior section.

Oddly, these issues only exist on Safari x VoiceOver x React Aria. 

We also tested on:
- [React spectrum](https://react-spectrum.adobe.com/react-spectrum/ActionMenu.html#actionmenu) ✅ 
- VoiceOver x Chrome (typically unrecommended, but worked 😅 )

We plan on testing on JAWs x Chrome once we procure a JAWs license too.

### 🤔 Expected Behavior?

Ideally, this should perform similarly to how it works on Chrome:
- Navigating between menu items should work smoothly without unexpected breaks, interruptions, or traps
- The first menu item should be announced when the menu is opened
- The menu should accurately count the number of menu items available

### 😯 Current Behavior

See above and replication steps 🙇 

### 💁 Possible Solution

This _may_ be a bug specific to Safari and VoiceOver, given these issues only emerge in this particular use case. Otherwise, I'm not sure how to begin debugging/fixing this 🙇 

### 🔦 Context

We're currently testing working on releasing a design system that is compatible with popular screen reader and browser combinations. Safari x VoiceOver is among the combinations we are hoping to be able to release to.

### 🖥️ Steps to Reproduce

**Using VoiceOver x Safari conduct the following tests**
- [Navigate to the first example on this page](https://react-spectrum.adobe.com/react-aria/examples/action-menu.html#action-menu)
- Open it using `VO` (typically `ctrl` `opt` ) and `space`
- Confirm the first menu item is not announced
  - If you can't open the menu with the `VO` command, try using `space` or `enter` to force it open
- Confirm the menu announces it has 7 items, when it only has 5 
  - This may be including the heading rules unintentionally, but I'm not 100% sure
- Attempt to navigate up and down the menu using `VO` and `Left/Right Arrow Keys`
- Confirm this is difficult
  - If you get stuck, you can try to force enter the menu using `Shift` `VO` `Down Arrow Key`
- Using `VO` and `Left Arrow Key`, navigate to the top until you can't go any further
  - You should hear "Dismiss menu"
- Then, using `VO` and `Right Arrow Key`, attempt to navigate downwards
  - You should hear "New..."
  - When attempting to continue, you should then hear "Dismiss"
  - You should then be unable to proceed unless reversing using `VO` and the `Left Arrow Key`

Thank you @snowystinger and @yihuiliao for correcting me on the arrow keys! 

### Version

React Aria

### What browsers are you seeing the problem on?

Safari

### What operating system are you using?

Mac OS

### 🧢 Your Company/Team

Northwestern Mutual

**Comment by yihuiliao** [comment]
I can confirm that when activating the button with  VO + space, the first item of the menu is not announced. However, if you use 'Enter' or 'Space' which are also valid ways of activating the button according to [Deque](https://media.dequeuniversity.com/en/docs/screen-readers-and-input-methods/voiceover-macos-guide.pdf), then the first item is announced. Seems weird that it works with some controls and not others but that might be tricky to debug. 

I think the reason why the menu is announcing 7 items instead of 5 is because of the separators in the examples. In Safari, when you navigate between the items, they are announced by the screenreader. If it's a purely visual thing, it may be possible to hide them by adding aria-hidden to the separators. In Chrome, it seems that these separators are skipped and never announced, hence why it's announcing 5 items. 

As for getting trapped, I was not able to reproduce that. What version of MacOS and Safari are you on? I'm on MacOS 15.2, Safari 18.2. Perhaps someone else on my team can try to reproduce the focus trap since that does not seem good.

**Comment by snowystinger** [comment]
I see what I think is the "getting trapped" you're talking about. It's getting stuck on the trigger, your unable to get into the menu?
I'm on MacOS 26.0.1

It looks like you are expected to use the Shift key to enter the Menu, which i confirmed did work for me.

I am not entirely sure why the VO cursor didn't immediately move into the Menu, it should have followed focus, this may be a VO bug.

**Comment by joshkimux** [comment]
@snowystinger that's one of the issues for sure! The other is when you get into the menu and attempt to navigate to the top of it and then back down.

I've attempted to record, but VO bugs up with screen recording on so it's been hard to get a clean example of it.

If it's helpful, I think the issue might be isolated just to react aria. It seems to work OK on react spectrum when I've tested.

**Comment by snowystinger** [comment]
@joshkimux I went back and read your instructions

> Using VO and Up Arrow Key, navigate to the top until you can't go any further
Then, using VO and Down Arrow Key, attempt to navigate downwards

 and realised I was navigating the menu with VO + Left/Right as opposed to up down. I think that's why I wasn't getting stuck. From the deque link that @yihuiliao shared, I don't see any VO + up/down commands.

**Comment by joshkimux** [comment]
@snowystinger thank you so much! My brain didn't connect the two together until too late. I'll update the replication steps. It should be VO Left/Right. My potato brain was using VO Left/Right but wrote it out as Up/Down 🤦

**Comment by joshkimux** [comment]
To better describe the trap, it specifically occurs when you reach the top of the menu and hear the `Dismiss` menu item.

Then, using `VO` and `Right Arrow Key`, attempt to navigate downwards
- You should hear "New..."
- When attempting to continue, you should then hear "Dismiss"
- You should then be unable to proceed unless reversing using `VO` and the `Left Arrow Key`

Expected behavior for navigating downwards should be:
- New -> Open -> etc

I'm unsure why Dismiss gets triggered again in the order after New

**Comment by snowystinger** [comment]
I'm unfortunately unable to reproduce the trap you're describing.
What version of Mac OS and Safari are you on? Are you running any extensions in Safari? Can you turn those off?

**Comment by joshkimux** [comment]
Apologies on the late reply friend!

> What version of Mac OS and Safari are you on? 

I'm on Sequoia 15.6.1 🫡 

> Are you running any extensions in Safari? Can you turn those off?

I'm not running any extensions in Safari 🙇 

I was also finally able to get a successful recording by turning on VoiceOver then recording (as opposed to the other way around). 

https://github.com/user-attachments/assets/c2ad0bea-bf66-4e78-823f-fb11cb0fa7cc

You'll notice the trap almost immediately once I enter the menu using `VO` `Shift` `Down Arrow` . I attempt to move forwards using `VO` `Right Arrow` , but am forced to use `VO` `Left Arrow` to move in reverse as I get caught on the dismiss button.

**Comment by yihuiliao** [comment]
Ah, I think I see what's happening. Basically the menu implementation is set up like this

```
<DismissButton>
<Menu />
<DismissButton>
```

So when you do the following:
> ...it specifically occurs when you reach the top of the menu and hear the Dismiss menu item.
>
> Then, using VO and Right Arrow Key, attempt to navigate downwards
> - You should hear "New..."
> - When attempting to continue, you should then hear "Dismiss"
> - You should then be unable to proceed unless reversing using VO and the Left Arrow Key

You are at the first dismiss button, then VO focuses on the menu as a whole, and then when you press the Right Arrow Key again, it goes to the second dismiss button, skipping the items inside the menu. Its behavior is confusing though since I'd expect it to enter the menu. I'm not sure why it's doing that.

The behavior in Chrome seems to make more sense. If you're at the top on the first dismiss button and then press the right arrow key, it'll announce "menu five items". Then if you press the right arrow key again, it'll enter the menu and focus the first item. Once you reach the last item, if you press the right arrow key, it'll say "end of menu", and if you press it again, you'll end up on the last dismiss button.

**Comment by yihuiliao** [comment]
Going to mark this as not planned since the concerns are mainly due to confusion about the VO controls. I don't want to entirely dismiss your concerns and I recognize that the VO experience is by no means perfect but I don't think it's necessarily an issue with our library and more so an issue with the screen readers in general.
